twake/backend/node/README.md

```diff
@@ -190,6 +190,65 @@ interface AdaptersConfiguration {
 }
 ```
 
+### Realtime CRUD Services
+
+The framework provides simple way to create CRUD Services which will notify clients using Websockets by following some conventions:
+
+1. The service must implement the `CRUDService` generic interface
+2. In order to push notification to clients, Typescript decorators must be added on methods (Create, Update, Delete methods)
+
+For example, let's say that we want to implement a CRUD Service for `messages`:
+
+```js
+import { CRUDService, CreateResult, DeleteResult, UpdateResult, EntityId } from "@/core/platform/framework/api/crud-service";
+
+class Message {
+  text: string;
+  createdAt: Date;
+  author: string;
+}
+
+class MessageService implements CRUDService<Message> {
+  async create(item: Message): Promise<CreateResult<Message>> {
+    // save the message then return a CreateResult instance
+  }
+
+  async get(id: EntityId): Promise<Message> {
+    // get the message from its ID and return it
+  }
+
+  async update(id: EntityId, item: Message): Promise<UpdateResult<Message>> {
+    // update the message with the given id, patch it with `item` values
+    // then return an instance of UpdateResult
+  }
+
+  async delete(id: EntityId): Promise<DeleteResult<Message>> {
+    // delete the message from its id then return a DeleteResult instance
+  }
+
+  async list(): Promise<Message[]> {
+    // get a list of messages
+  }
+}
+```
+
+By implementing the CRUD service following the API, we can now add realtime message to clients connected to Websocket on a selected collection of CRUD operations. For example, if we want to just notify when a `message` is created, we just have to add the right decorator on the `create` method like:
+
+```js
+import { RealtimeCreated } from "@/core/platform/framework/decorators";
+
+// top and bottom code removed for clarity
+class MessageService implements CRUDService<Message> {
+  @RealtimeCreated<Message>("/messages")
+  async create(item: Message): Promise<CreateResult<Message>> {
+    // save the message then return a CreateResult instance
+  }
+  // ...
+}
+```
+
+The `RealtimeCreated` decorator will intercept the `create` call and will publish an event in an internal event bus with the creation result, the input data and the `"/messages"` path. On the other side of the event bus, an event listener will be in charge of delivering the event to the right Websocket clients.
+
```

Comment:
@RomaricMourgues All that you describe here will be handled in the next PRs as we already discussed.

Also, can you describe what you expect for the `@RealtimeListed` decorator?

Reply:
I expect JWT token to be sent after a listing if we used the ?websocket=1 query parameter.
- When a listing is done we know that the user has access to this data,
- We generate JWT tokens containing `user_id` and `websocket_room` for each room we allow the user to be in (can be multiple routes in the channel example),
- The JWT will have a lifetime of 1-5 minutes
- This JWT token will be used on the frontend to join the websocket room,
- And then the websocket server will just read the JWT token and will know the user is allowed to join this specific room.

On a reconnect after a websocket disconnection, we use again a REST http list call to obtain a new JWT token and join the websocekt room.

The idea is to have zero access verification on the websocket server except JWT. It will allow us to have:
- a faster backend
- no code duplication (is we have access to the list, then we have access to the websocket rooms)

Reply:
Nice idea, will think about it 💪🏼

Reply:
Great 😄